sky/skylet/providers/ibm/node_provider.py

```diff
@@ -0,0 +1,965 @@
+                        )
+            # dump in-memory cache to local cache (file).
+            self.set_node_tags(None, None)
+
+        else:
+            name = socket.gethostname()  # returns the instance's (VSI) name
+            logger.debug(f"Check if {name} is HEAD")
+            # check if current VM is the remote head node
+            if self._get_node_type(name) == NODE_KIND_HEAD:
+                logger.debug(f"{name} is HEAD")
+                # pylint: disable=line-too-long
+                node = self.ibm_vpc_client.list_instances(name=name).get_result()[
+                    "instances"
+                ]
+                if node:
+                    logger.debug(f"{name} is node in vpc")
+
+                    # remote head node will calculate 2 hash values below
+                    ray_bootstrap_config = Path.home() / "ray_bootstrap_config.yaml"
+                    config = json.loads(ray_bootstrap_config.read_text())
+                    (runtime_hash, mounts_contents_hash) = hash_runtime_conf(
+                        config["file_mounts"], None, config
+                    )
+
+                    # launch_hash is used to verify the head node belongs
+                    # to the cluster. e.g. checked when running --restart-only
+                    head_node_config = {}
+                    head_node_type = config.get("head_node_type")
+                    if head_node_type:
+                        head_config = config["available_node_types"][head_node_type]
+                        head_node_config.update(head_config["node_config"])
+                    launch_hash = hash_launch_conf(head_node_config, config["auth"])
+
+                    head_tags = {
+                        TAG_RAY_NODE_KIND: NODE_KIND_HEAD,
+                        TAG_RAY_NODE_NAME: name,
+                        TAG_RAY_NODE_STATUS: "up-to-date",
+                        TAG_RAY_CLUSTER_NAME: self.cluster_name,
+                        TAG_RAY_USER_NODE_TYPE: config["head_node_type"],
+                        TAG_RAY_RUNTIME_CONFIG: runtime_hash,
+                        TAG_RAY_LAUNCH_CONFIG: launch_hash,
+                        TAG_RAY_FILE_MOUNTS_CONTENTS: mounts_contents_hash,
+                    }
+
+                    logger.debug(f"Setting HEAD node tags {head_tags}")
+                    self.set_node_tags(node[0]["id"], head_tags)
+
+    def __init__(self, provider_config, cluster_name):
+        """
+        Args:
+            provider_config (dict): containing the provider segment of
+                the cluster's config file.initialized as an instance
+                variable of the parent class, hence can accessed via self.
+            cluster_name(str): value of cluster_name within the
+                 cluster's config file.
+
+        """
+        NodeProvider.__init__(self, provider_config, cluster_name)
+
+        self.lock = threading.RLock()
+        self.tags_file = Path.home() / VPC_TAGS
```

Comment:
Is it possible to have race condition if all clusters share the same tag file (e.g. one cluster reads from file, another cluster writes to file, and then the first cluster writes to file).

Reply:
So each cluster will interact with a different segment of the file, but it still has to load the entire file, right? So if X loads the entire file and changes entry 1, Y loads the entire file and changes entry 2, X writes back, Y writes back, then X's changes will not persist.

Reply:
A lock is set on writers to the shared resource [here](https://github.com/skypilot-org/skypilot/blob/d41e38a22ad189bc1b791bd655c0256044778680/sky/skylet/providers/ibm/node_provider.py#L483).
You are completely right, had we didn't have a lock on the file, it would be classified as a classic race condition.